rfcs/003 - Multi-Region Support RFC.md

```diff
@@ -0,0 +1,161 @@
+
+In order to keep the current flexibility, there would be a Postgres function provided by the `aws_account` module to return the default region they have selected. This could simply be a rename of the current `region` column on the `aws_account` table to `default_region`, but we also want to improve on the current state of things when moving to multi-region modules.
+
+The first thing that we desire is that the default region be guaranteed to be an actual region that the user is able to access. That depends on their credentials, so we should have an `aws_regions` table contains the list of all regions that we acquire via the API, and then joined in on the `default_region` column.
+
+The second thing that is necessary with this approach is that the default region function output needs to always return a region.
+
+These two are not compatible with each other, and here's why:
+
+If we put a foreign key join on the `aws_account` table to the `aws_regions` table, then when we first attempt to insert the credentials to make it available to the Mapper in charge of the `aws_regions` table, we would have to have the `default_region` be set to `NULL` to not fall afoul of the foreign key constraints. But if it is `NULL` then the default value function would need to have some other value it always spits out to users which we don't *know* for sure could be something like `us-east-1` because it depends on the response for the regions available to the credentials in question.
+
+There's also a third desire for this refactoring: we want to go multi-region, but we also want to allow users to disable regions that they do not use (or do not want to manage with IaSQL) to improve the performance of `apply` and `sync` operations.
+
+I bring this up because it suggests a solution for both problems.
+
+We split the `aws_account` into two tables, `aws_credentials` and `aws_regions`. The credentials table *only* has the credentials for the AWS account, while the `aws_regions` table has three columns, the `region`, `is_default`, and `is_enabled`, with the last two being boolean columns and the `is_default` column having a constraint that only one row has a default value.
+
+When first populating the `aws_regions` table, all regions returned that the user has access to should be enabled by default, and the first one is set as the default. We can then overload the `aws_default_region` function we intend to use for reading the default region to also perform the necessary operations to alter the default region (this requires temporarily disabling the constraint on the table only allowing one `is_default == true` row, making the current default `false`, then setting the new default to `true` and re-enabling the constraint within a transaction).
```

Comment:
This means that the `aws_account` mapper would be able to make schema changes, right? That would be the first mapper of its kind?

Reply:
No? The schema would stay the same, calling the `aws_default_region()` function without arguments to return the default region as the default value if not provided on `insert` in other schemas.

The schema these other modules would use would have a dynamic default value instead of a static one, but the schema itself is static.

Reply:
Got it. I guess it within the function and not the mapper. LGTM.